sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+
+    def create_encoder_layer(self, input_num_hidden: int, prefix: str) -> EncoderLayer:
+        # TODO: can we simplify this? (e.g. by using all attributes of the config object)
+        return MultiHeadSelfAttentionEncoderLayer(num_hidden=self.num_hidden,
+                                                  att_num_hidden=self.att_num_hidden,
+                                                  dropout=self.dropout,
+                                                  dropout_attention=self.dropout_attention,
+                                                  heads=self.heads,
+                                                  prefix=prefix + "mh_self_att_")
+
+    def create_decoder_layer(self, input_num_hidden: int, prefix: str) -> DecoderLayer:
+        return MultiHeadSelfAttentionDecoderLayer(num_hidden=self.num_hidden,
+                                                  att_num_hidden=self.att_num_hidden,
+                                                  dropout=self.dropout,
+                                                  dropout_attention=self.dropout_attention,
+                                                  heads=self.heads,
+                                                  prefix=prefix + "mh_self_att_")
+
+
+# TODO: make sure the number of hidden units does not change!
+class ResidualEncoderLayer(EncoderLayer):
+    def __init__(self, layers: List[EncoderLayer]):
+        self.layers = layers
+
+    def encode_sequence(self,
+                        source_encoded: mx.sym.Symbol,
+                        source_encoded_lengths: mx.sym.Symbol,
+                        source_encoded_max_length: int,
+                        att_dict: Optional[Dict[str, mx.sym.Symbol]]) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
+        new_source_encoded = source_encoded
+        for layer in self.layers:
+            new_source_encoded, new_source_encoded_lengths, new_source_encoded_max_length = layer.encode_sequence(
+                new_source_encoded,
+                source_encoded_lengths,
+                source_encoded_max_length,
+                att_dict)
+            assert source_encoded_max_length == new_source_encoded_max_length, C.SEQUENCE_LENGTH_MUST_NOT_CHANGE_MSG
+            assert source_encoded_lengths is new_source_encoded_lengths, C.SEQUENCE_LENGTH_MUST_NOT_CHANGE_MSG
+
+        return source_encoded + new_source_encoded, source_encoded_lengths, source_encoded_max_length
+
+    def att_names(self):
+        att_names = []
+        for layer in self.layers:
+            att_names.extend(layer.att_names())
+        return att_names
+
+    def get_num_hidden(self) -> int:
+        return self.layers[-1].get_num_hidden()
+
+
+# TODO: potentially add a projection layer (for when the shapes don't match up). Alternative: check that the input num hidden matches the output num_hidden (maybe add a get_input_num_hidden())
+class ResidualDecoderLayer(NestedDecoderLayer):
```

Comment:
why can't we have a ResidualLayer that inherits from SharedEncoderDecoderLayer and implements all 3 methods (encode_sequence, decode_sequence, decode_step)?

Reply:
you mean inheriting from both NestedDecoderLayer and SharedEncoderDecoderLayer? yes, we potentially could. I'll add a TODO.